src/Compilers/CSharp/Portable/Utilities/ValueSetFactory.SingleTC.cs

```diff
@@ -0,0 +1,104 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+
+#nullable enable
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    using static BinaryOperatorKind;
+
+    internal static partial class ValueSetFactory
+    {
+        private struct SingleTC : FloatingTC<float>, NumericTC<float>
+        {
+            float NumericTC<float>.MinValue => float.MinValue;
+
+            float NumericTC<float>.MaxValue => float.MaxValue;
+
+            float FloatingTC<float>.NaN => float.NaN;
+
+            float FloatingTC<float>.MinusInf => float.NegativeInfinity;
+
+            float FloatingTC<float>.PlusInf => float.PositiveInfinity;
+
+            float NumericTC<float>.Next(float value)
+            {
+                Debug.Assert(!float.IsNaN(value));
+                Debug.Assert(!float.IsInfinity(value));
+                Debug.Assert(value != float.MaxValue);
+                if (value < 0)
+                {
+                    if (value == -float.Epsilon)
```

Comment:
>if (value == -float.Epsilon) [](start = 20, length = 28)

Can we add C# tests for switch completeness involving negative zero and zero? What is the expected behavior (they are equivalent?)?

Reply:
The input value to `Next` always comes from the `Split` function, which will never produce negative zero.  There is no way to write a negative zero double constant in C#, so there is no way to write the test you suggest.  On the other hand, a computed *input* value of negative zero is indistinguishble from zero, since all we do is essentially emit the comparisons that appeared in source.

---
In reply to: [381438744](https://github.com/dotnet/roslyn/pull/41503#discussion_r381438744) [](ancestors = 381438744)